torch/nn/parallel/distributed.py

```diff
@@ -142,23 +142,35 @@ def forward(ctx, reducer, state_dict, *inputs):
     @staticmethod
     def backward(ctx, *grad_outputs):
         state_dict = ctx.state_dict
+
+        grad_enabled = state_dict['grad_enabled']
+        require_backward_grad_sync = state_dict['require_backward_grad_sync']
         static_graph_training = ctx.state_dict['static_graph']
-        if static_graph_training and ctx.state_dict['num_iterations'] == 1:
+        if grad_enabled and require_backward_grad_sync:
+            if static_graph_training or not state_dict['find_unused']:
+                ctx.reducer.prepare_for_backward([])
+            else:
+                # First type of unused params: parameters that did not participate
+                # in computing model outputs. These are found by the below call to
+                # prepare_for_backward.
+                # Second type of unused params: params that won't get gradient
+                # because outputs they produced do not get used in computing loss
+                # for this call to backward. Due to this passthrough autograd
+                # function, autograd hooks for these parameters are now triggered
+                # with undefined gradient to maintain parity with local training.
+                # DDP takes care of undefined grads in this case to ensure the .grad
+                # field of the param is not touched.
+                ctx.reducer.prepare_for_backward(list(_find_tensors(ctx.inputs)))
+
+        # Note that we enqueue delay allreduce after prepare_for_backward in
+        # static graph training as prepare_for_backward sets the
+        # num_backwards_call counter in the reducer.
+        static_graph_first_bwd = (
+            static_graph_training and ctx.reducer._static_graph_first_bwd()
```

Comment:
nit: I'm wondering why static_graph_training is needed here? _static_graph_first_bwd() has already checked this?

Reply:
If we are not in static graph mode at all, it would save one Python -> C++ -> Python round trip. I feel like that might be worth it as a small optimization in non-static graph mode.